### PR TITLE
tests: lib: gui: lvgl: add LittleFS block device support

### LINKED ISSUE #67685: Tests: lib: gui: lvgl: test causes a Kernel Panic on stm32h747i_disco_m7
**Describe the bug**
Tests/lib/gui/lvgl test does not run on stm32h747i_disco_m7, & possibly other platforms as well, & causes a **kernel panic**. LittleFS appears to be the cause (as seen in the console output below).

 - target platform: tested on stm32h747i_disco_m7
 - Attempted workaround: update LittleFS to 2.8.1 using [PR 65414 ](https://github.com/zephyrproject-rtos/zephyr/pull/65414)
 - Still don't know if it's a regression

**To Reproduce**
1. cd ~/zephyrproject/zephyr
2. west build -p always -b stm32h747i_disco_m7 tests/lib/gui/lvgl -DSHIELD="st_b_lcd40_dsi1_mb1166"
4. west flash

**Expected behavior**
Random noise on the screen

**Impact**
Cannot run test

**Logs and console output**

```
I: Reading SFDP
I: qspi-nor-flash-1@0: SFDP v 1.6 AP ff with 2 PH
I: PH0: ff00 rev 1.6: 16 DW @ 30
I: Reading SFDP
I: qspi-nor-flash-1@0: 64 MiBy flash
I: Flash - address mode: 4B
I: Quad read mode 4 instr [0x6b] supported
I: Quad read mode 7 instr [0xeb] supported
I: Quad read mode 7 instr [0xeb] will be used
I: QE requirement mode: 0
I: Quad mode enabled
I: PH1: ff84 rev 1.0: 2 DW @ 80
*** Booting Zephyr OS build zephyr-v3.5.0-4181-g4204ca9bcb3f ***
Running TESTSUITE lvgl_fs
===================================================================
I: LittleFS version 2.5, disk version 2.0
I: FS at flash-controller@52002000:0xff800 is 0 0x20000-byte blocks with 512 cycle
I: sizes: rd 16 ; pr 16 ; ca 64 ; la 32
ASSERTION FAIL [block_count != 0] @ WEST_TOPDIR/zephyr/subsys/fs/littlefs_fs.c:813
E: r0/a1:  0x00000004  r1/a2:  0x0000032d  r2/a3:  0x40011000
E: r3/a4:  0x08037ec9 r12/ip:  0x00000000 r14/lr:  0x0800276f
E:  xpsr:  0x61000000
E: Faulting instruction address (r15/pc): 0x080270e4
E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
E: Current thread: 0x24004670 (main)
E: Halting system
```

**Environment:**
 - OS: Linux
 - Toolchain: Zephyr SDK 0.16.3
 - Commit SHA: main branch, 4204ca9bcb3f675a0e63f647931ffe2c4b515729

**Additional context**
None for now

### LINKED ISSUE #67686: Tests: lib: gui: lvgl: test causes a Stack Overflow on stm32f746g_disco
**Describe the bug**
Tests/lib/gui/lvgl test does not run on stm32f746g_disco, & possibly other platforms as well, & causes a Stack Overflow.

 - target platform: tested on stm32f746g_disco
 - Attempted workaround: update LittleFS to 2.8.1 using [PR 65414 ](https://github.com/zephyrproject-rtos/zephyr/pull/65414)
 - Still don't know if it's a regression

**To Reproduce**
1. cd ~/zephyrproject/zephyr
2. west build -p always -b stm32f746g_disco tests/lib/gui/lvgl
3. west flash

**Expected behavior**
Random noise on the screen

**Impact**
Cannot run test

**Logs and console output**

```
I: Reading SFDP
I: qspi-nor-flash@0: SFDP v 1.5 AP ff with 2 PH
I: PH0: ff00 rev 1.5: 16 DW @ 30
I: Reading SFDP
I: qspi-nor-flash@0: 16 MiBy flash
I: PH1: ff03 rev 1.0: 2 DW @ 100
* Booting Zephyr OS build zephyr-v3.5.0-4181-g4204ca9bcb3f *
Running TESTSUITE lvgl_fs
===================================================================
I: LittleFS version 2.5, disk version 2.0
I: FS at qspi-nor-flash@0:0xa0000 is 3840 0x1000-byte blocks with 512 cycle
I: sizes: rd 16 ; pr 16 ; ca 64 ; la 32
E: * MPU FAULT *
E:   Stacking error (context area might be not valid)
E:   Data Access Violation
E:   MMFAR Address: 0x2001da38
E: r0/a1:  0x00000002  r1/a2:  0x2001da5c  r2/a3:  0x2001da5c
E: r3/a4:  0x00000000 r12/ip:  0x00000000 r14/lr:  0x2001da70
E:  xpsr:  0x00000000
E: Faulting instruction address (r15/pc): 0x2001da70
E: >>> ZEPHYR FATAL ERROR 2: Stack overflow on CPU 0
E: Current thread: 0x200146f0 (main)
E: Halting system
```

**Environment:**
 - OS: Linux
 - Toolchain: Zephyr SDK 0.16.3
 - Commit SHA: main branch, 4204ca9bcb3f675a0e63f647931ffe2c4b515729

**Additional context**
None for now

### PR DESCRIPTION
**Pull request by ajarmouni-st**
Add support for running LVGL test, that uses LittleFS, on a block device (MMC/SDMMC). Currently, only Flash is supported.

### CODE REVIEW

**Review by faxe1008 — COMMENTED** [review]
Can you maybe add  a testcase to ensure that this is infact built in CI?

File: tests/lib/gui/lvgl/src/main.c [hunk 1]
```diff
@@ -71,7 +71,9 @@ ZTEST(lvgl_screen, test_add_delete_screen)
 
 	lv_scr_load(new_screen);
 
+#ifndef CONFIG_SOC_FAMILY_STM32
```
Comment (faxe1008) [inline tests/lib/gui/lvgl/src/main.c:74]: Can you explain why this is needed?

File: tests/lib/gui/lvgl/boards/stm32h747i_disco_m7.overlay [hunk 2]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) 2024 STMicroelectronics
```
Comment (nordicjm) [inline tests/lib/gui/lvgl/boards/stm32h747i_disco_m7.overlay:2]: Check that board names are still valid with hwmv2

Comment (ajarmouni-st) [reply tests/lib/gui/lvgl/boards/stm32h747i_disco_m7.overlay]: done

File: tests/lib/gui/lvgl/src/main.c (hunk 1 above)
Comment (faxe1008) [reply tests/lib/gui/lvgl/src/main.c:74]: I just ran the test on my mimxrt1060_evkb and the issue is also present on there. From my point of view there are two possibilities:

- Find out what the issue is and possibly submit upstream patch (I suspect that something is going wrong in the animation code.
- Remove the handler call completely. If we are sure that `lv_img_set_src` will call through to the decoder/fs layer even without the handler I am fine with removing the handler. For the screen creation we are not using an animation, therefore they should not be needed.

Comment (ajarmouni-st) [reply tests/lib/gui/lvgl/src/main.c:74]: I am inclined to follow the second route.

Comment (faxe1008) [reply tests/lib/gui/lvgl/src/main.c]: @ajarmouni-st 
Today I tried fiddeling with various `STACK_SIZE` settings but I could not pin point the issue further. Stacks that I checked:
```
CONFIG_ISR_STACK_SIZE=4096
CONFIG_INPUT_THREAD_STACK_SIZE=4096
CONFIG_ZTEST_STACK_SIZE=4096
CONFIG_LV_Z_FLUSH_THREAD=n
```
I also tried increasing `CONFIG_LV_Z_MEM_POOL_SIZE`, but that did not matter either. Need to do some further tests tomorrow.

Comment (ajarmouni-st) [reply tests/lib/gui/lvgl/src/main.c]: Hi @faxe1008 thanks for taking the time to investigate this. I also tried to increase all kind of stack sizes but it had no effect whatsoever.

Comment (faxe1008) [reply tests/lib/gui/lvgl/src/main.c]: I verified that the decoder is accessing the fs implementation even without the call to the handler:
```
#0  lvgl_fs_read (drv=0x80a2a00 <fs_drv>, file=0x80a3a64 <lvgl_heap_mem+3684>, buf=0xf7c782b4, btr=4, br=0xf7c781e4) at /zephyr/modules/lvgl/lvgl_fs.c:99
#1  0x0807e752 in lv_fs_read (file_p=0xf7c78230, buf=0xf7c782b4, btr=4, br=0xf7c7822c) at /modules/lib/gui/lvgl/src/misc/lv_fs.c:218
#2  0x08065197 in lv_img_decoder_built_in_info (decoder=0x80a2c6c <lvgl_heap_mem+108>, src=0x8097b39, header=0xf7c782b4)
    at /modules/lib/gui/lvgl/src/draw/lv_img_decoder.c:294
#3  0x08065bd0 in lv_img_decoder_get_info (src=0x8097b39, header=0xf7c782b4) at /modules/lib/gui/lvgl/src/draw/lv_img_decoder.c:99
#4  0x08086741 in lv_img_set_src (obj=0x80a3a04 <lvgl_heap_mem+3588>, src=0x8097b39) at /modules/lib/gui/lvgl/src/widgets/lv_img.c:107
```
So removing this one should be ok - however I am intruiged to find out why this happens 😄

Comment (ajarmouni-st) [reply tests/lib/gui/lvgl/src/main.c]: > I verified that the decoder is accessing the fs implementation even without the call to the handler

Great, Thanks!

> however I am intruiged to find out why this happens

Me too, but it appears to be really convoluted (I spent a considerable time investigating this before finding that commenting the handler calls solves the issue, & that's when I gave up ^ ^)

**Comment by ajarmouni-st** [comment]
Rebased to resolve conflicts.

**Comment by ajarmouni-st** [comment]
Hi @faxe1008, should we proceed with the PR in its current status?

**Review by faxe1008 — APPROVED** [review]
@ajarmouni-st I thought that maybe I find the issue, but since I am currently working on the v9.1 migration of the module the tests need to be adjusted anyways, so fine with this going in.

**Review by erwango — APPROVED** [review]

**Pull request merged**